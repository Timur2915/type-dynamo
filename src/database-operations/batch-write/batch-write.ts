import { DynamoDB } from 'aws-sdk'
import DynamoToPromise from '../dynamo-to-promise'
import { deepClone } from '../helpers'
export interface BatchWriteResult<TableModel> {
    data: TableModel[]
}

const segmentBatchWriteInputs = (batchWriteInput: DynamoDB.BatchWriteItemInput) => {
    const RequestItems = deepClone(batchWriteInput.RequestItems)
    const tableName = Object.keys(RequestItems)[0]
    const segments = new Array<DynamoDB.BatchWriteItemInput>()
    do {
        const segment = deepClone(batchWriteInput)
        segment.RequestItems[tableName] = deepClone(RequestItems[tableName].slice(0, 25)),
        segments.push(segment)
        if (RequestItems[tableName].length > 25) {
            RequestItems[tableName] = RequestItems[tableName].slice(25, RequestItems[tableName].length)
        } else {
            RequestItems[tableName] = []
        }
    } while (RequestItems[tableName].length)
    return segments
}

const singleBatchWrite = async <Entity> (
    batchWriteInput: DynamoDB.BatchWriteItemInput,
    dynamoPromise: DynamoToPromise,
) => {
    let batchWriteOutput = await dynamoPromise.batchWrite(batchWriteInput)
    while (batchWriteOutput.UnprocessedItems && Object.keys(batchWriteOutput.UnprocessedItems).length) {
        batchWriteInput.RequestItems = batchWriteOutput.UnprocessedItems
        batchWriteOutput = await dynamoPromise.batchWrite(batchWriteInput)
    }
}

export async function batchWrite<
    Entity
>(
    items: Entity[], batchWriteInput: DynamoDB.BatchWriteItemInput, dynamoPromise: DynamoToPromise,
): Promise<BatchWriteResult<Entity>> {
    const segmentedBatchWriteInputs = segmentBatchWriteInputs(batchWriteInput)
    const batchWrites = segmentedBatchWriteInputs.map( (segment) => singleBatchWrite(
        segment, dynamoPromise,
    ))
    await Promise.all(batchWrites)
    const result: BatchWriteResult<Entity> = {
        data: items,
    }
    return result
}
