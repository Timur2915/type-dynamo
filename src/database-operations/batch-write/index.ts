export * from './build-batch-write-input'
export * from './batch-write'
