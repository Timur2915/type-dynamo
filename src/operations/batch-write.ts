import { DynamoDB } from 'aws-sdk'
import { BatchWriteItemInput } from 'aws-sdk/clients/dynamodb'
import { IBatchWriteInput, IBatchWriteResult, IHelpers } from '../types'

export class BatchWrite {
  private MAX_ITEMS_PER_BATCH = 25
  private INITIAL_BACKOFF_TIME = 100 // ms
  private MAX_BACKOFF_TIME = 1000 // ms

  constructor(
    private dynamoClient: DynamoDB, private helpers: IHelpers,
  ) { }

  public async execute(
    input: IBatchWriteInput<any>,
  ): Promise<IBatchWriteResult<any>> {
    const dynamoInput = this.buildInputFrom(input)
    const chunks = this.splitInputIntoChunks(dynamoInput, input.tableName)
    for (const chunk of chunks) {
      await this.batchWriteItem(chunk)
    }
    return {data: input.items}
  }

  private buildInputFrom(input: IBatchWriteInput<any>) {
    return {
      RequestItems: {
        [input.tableName]: input.items.map((item) => ({
          PutRequest: {Item: DynamoDB.Converter.marshall(item)},
        })),
      },
    }
  }

  private async batchWriteItem(input: BatchWriteItemInput) {
    let unprocessedItems: DynamoDB.BatchWriteItemRequestMap = {}
    let backoffTime = this.INITIAL_BACKOFF_TIME
    let requestFailed = false
    do {
      if (Object.keys(unprocessedItems).length) {
        input = {...input, RequestItems: unprocessedItems}
      }
      try {
        const {
          UnprocessedItems,
        } = await this.dynamoClient.batchWriteItem(input).promise()
        unprocessedItems = UnprocessedItems || {}
        requestFailed = false
      } catch (err) {
        requestFailed = true
        switch (err.code) {
          case 'ProvisionedThroughputExceededException': {
            await this.helpers.timeout.wait(backoffTime)
            backoffTime *= 2
            break
          }
          default:
            throw err
        }
      }
    } while (
        Object.keys(unprocessedItems).length  ||
        (requestFailed && backoffTime < this.MAX_BACKOFF_TIME)
    )
  }

  private splitInputIntoChunks(
    input: BatchWriteItemInput, tableName: string,
  ): BatchWriteItemInput[] {
    let chunk = 0
    return input.RequestItems[tableName].reduce((acc, value, index) => {
      if (index % this.MAX_ITEMS_PER_BATCH === 0 && index !== 0) {
        chunk++
        acc[chunk] = {RequestItems: {[tableName]: []}}
      }
      acc[chunk].RequestItems[tableName] = [
        ...acc[chunk].RequestItems[tableName], value,
      ]
      return acc
    }, new Array<BatchWriteItemInput>({
      RequestItems: {[tableName]: []},
    }))
  }
}
