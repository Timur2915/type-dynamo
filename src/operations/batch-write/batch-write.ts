import { AWSError, DynamoDB } from 'aws-sdk'
import { merge } from 'lodash'
import DynamoClient from '../dynamo-client'
import { timeoutPromise } from '../helpers'
import { deepClone } from '../helpers'
const marshalItem = require('dynamodb-marshaler').marshalItem

const BATCH_WRITE_LIMIT = 25 // items
const INITIAL_BACKOFF = 50 // ms
const MAX_BACKOFF = 1000 // ms

export interface BatchWriteResult<TableModel> {
    data: TableModel[]
}

const segmentBatchWriteInputs = (batchWriteInput: DynamoDB.BatchWriteItemInput) => {
    const RequestItems = deepClone(batchWriteInput.RequestItems)
    const tableName = Object.keys(RequestItems)[0]
    const segments = new Array<DynamoDB.BatchWriteItemInput>()
    do {
        const segment = deepClone(batchWriteInput)
        segment.RequestItems[tableName] = RequestItems[tableName].slice(0, BATCH_WRITE_LIMIT)
        segments.push(segment)
        if (RequestItems[tableName].length > BATCH_WRITE_LIMIT) {
            RequestItems[tableName] = RequestItems[tableName].slice(
                BATCH_WRITE_LIMIT, RequestItems[tableName].length,
            )
        } else {
            RequestItems[tableName] = []
        }
    } while (RequestItems[tableName].length)
    return segments
}

const singleBatchWrite = async <Entity> (
    batchWriteInput: DynamoDB.BatchWriteItemInput,
    dynamoClient: DynamoClient,
) => {
    const tableName =  Object.keys(batchWriteInput.RequestItems)[0]
    try {
        const batchWriteOutput = await dynamoClient.batchWrite(batchWriteInput)
        if (batchWriteOutput.UnprocessedItems && Object.keys(batchWriteOutput.UnprocessedItems).length) {
            batchWriteInput.RequestItems[tableName] =
                batchWriteOutput.UnprocessedItems[tableName].map( (request) => ({
                    PutRequest: { Item: marshalItem(request.PutRequest!.Item) },
                }))
            return batchWriteInput
        }
        return undefined
    } catch (err) {
        if ((err as AWSError).code === 'ProvisionedThroughputExceededException') {
            return batchWriteInput
        } else {
            throw new Error('UnknownError')
        }
    }
}

const mergeArray = (arr1: any[], arr2: any[]) => {
    let arrMerge = new Array()
    arrMerge = arrMerge.concat(arr1)
    arrMerge = arrMerge.concat(arr2)
    return arrMerge
}

export async function batchWrite<
    Entity
>(
    items: Entity[], batchWriteInput: DynamoDB.BatchWriteItemInput, dynamoClient: DynamoClient,
): Promise<BatchWriteResult<Entity>> {
    let unprocessed = batchWriteInput
    const tableName = Object.keys(unprocessed.RequestItems)[0]
    do {
        const segments = segmentBatchWriteInputs(unprocessed)
        const batchWrites = segments.map( (segment) => singleBatchWrite<Entity>(
            segment, dynamoClient,
        ))
        unprocessed = (await Promise.all(batchWrites)).reduce((acc, current) => {
            if (current) {
                acc.RequestItems[tableName] = mergeArray(acc.RequestItems[tableName], current.RequestItems[tableName])
            }
            return acc
        }, { RequestItems : { [tableName]: [] } } as any)
    } while (unprocessed.RequestItems[tableName].length)
    return { data: items }
}
