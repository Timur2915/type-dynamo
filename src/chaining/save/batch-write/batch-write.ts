import { BatchWriteResult } from '../../../database-operations/batch-write'
import { EntitySchema } from '../../../schema'
import { Chaining } from '../../common'
import { BatchWriteChainingKind } from './'
import { execute } from './execute'

export type BatchWriteType = 'batchWrite'

export interface BatchWrite<Table> {
    schema: EntitySchema,
    items: Table[],
}

export class DynamoBatchWrite<Entity> extends Chaining<BatchWriteChainingKind> {

    protected _batchWrite: BatchWrite<Entity>

    constructor(
        batchWrite: BatchWrite<Entity>,
    ) {
        super('batchWrite')
        this._batchWrite = batchWrite
        this._stack.push(this)
    }

    public execute() {
        return execute<Entity>(this._stack)
    }

}
