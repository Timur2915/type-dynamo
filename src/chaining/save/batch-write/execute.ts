import { DynamoDB } from 'aws-sdk'
import { batchWrite, buildBatchWriteInput} from '../../../database-operations/batch-write'
import { EntitySchema } from '../../../schema'
import { Chaining, Filter, Paginate, WithAttributes } from '../../common'
import { BatchWriteChainingKind } from './'
import { BatchWrite } from './batch-write'

function extractFromStack<Table>(stack: Array<Chaining<BatchWriteChainingKind>>): {
    batchWriteMetadata: BatchWrite<Table>,
} {
    const batchWriteMetadata = (stack[0] as any)._batchWrite
    return { batchWriteMetadata }
}

export function execute<Entity>(
    stack: Array<Chaining<BatchWriteChainingKind>>,
) {
    const { batchWriteMetadata } = extractFromStack<Entity>(stack)
    const batchWriteInput = buildBatchWriteInput(batchWriteMetadata)
    return batchWrite<Entity>(batchWriteMetadata.items, batchWriteInput)
}
