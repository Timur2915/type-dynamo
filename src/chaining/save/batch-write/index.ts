import { BatchWriteType } from './batch-write'

export type BatchWriteChainingKind = BatchWriteType

export { DynamoBatchWrite } from './batch-write'
