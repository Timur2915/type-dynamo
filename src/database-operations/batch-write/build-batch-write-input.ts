import { DynamoDB } from 'aws-sdk'
import { WithAttributes } from '../../chaining/common'
import { BatchWrite } from '../../chaining/save/batch-write/batch-write'
import { EntitySchema } from '../../schema'
import { buildKey, projectionExpression } from '../helpers'
const marshalItem = require('dynamodb-marshaler').marshalItem

export function buildBatchWriteInput<Table>(
    batchWrite: BatchWrite<Table>,
): DynamoDB.BatchWriteItemInput {
    const input: DynamoDB.BatchWriteItemInput = {
        RequestItems: {
            [batchWrite.schema.tableName]: [
                ...batchWrite.items.map( (item) => ({ PutRequest: { Item: marshalItem(item) } })),
            ],
        },
    }
    return input
}
