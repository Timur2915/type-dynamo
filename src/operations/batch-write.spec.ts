import { BatchWrite, IBatchWriteInput } from './batch-write'

describe('BatchWrite', () => {
  let dynamoClient: any
  let batchWrite: BatchWrite
  let input: IBatchWriteInput<any>

  beforeEach(() => {
    input = {
      tableName: 'DummyTable',
      items: [
        { id: '1', email: 'john@email.com' },
        { id: '2', email: 'doe@email.com' },
      ],
    }
    dynamoClient = {
      batchWriteItem: jest.fn(() => ({
        promise: async () => null,
      })),
    }
    batchWrite = new BatchWrite(dynamoClient as any)
  })

  it('calls dynamoClient correctly', async () => {
    await batchWrite.execute(input)
    expect(dynamoClient.batchWriteItem).toHaveBeenCalledWith({
      RequestItems: {
        DummyTable: [{
          PutRequest: {
            Item: { id: { S: '1' }, email: { S: 'john@email.com' } },
          },
        }, {
          PutRequest: {
            Item: { id: { S: '2' }, email: { S: 'doe@email.com' } },
          },
        }],
      },
    })
  })

  it('returns items in correct format', async () => {
    expect(await batchWrite.execute(input)).toEqual({
      data: [
        { id: '1', email: 'john@email.com' },
        { id: '2', email: 'doe@email.com' },
      ],
    })
  })

  // @ts-ignore
  context('when more than 25 items are requested', () => {
    const generateFakeItems = (total: number) => {
      const items = []
      for (let i = 0; i < total; i++) {
        items.push({ id: i.toString(), email: `${i}@email.com`})
      }
      return items
    }

    beforeEach(() => {
      input.items = generateFakeItems(77)
    })

    it('splits requests into chunks of max of 25 items each', async () => {
      await batchWrite.execute(input)
      expect(dynamoClient.batchWriteItem.mock.calls).toMatchSnapshot()
    })

    it('calls dynamoClient for each chunk', async () => {
      await batchWrite.execute(input)
      expect(dynamoClient.batchWriteItem).toHaveBeenCalledTimes(4)
    })
  })
})
