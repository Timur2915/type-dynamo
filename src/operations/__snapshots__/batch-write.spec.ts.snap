// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchWrite when more than 25 items are requested splits requests into chunks of max of 25 items each 1`] = `
Array [
  Array [
    Object {
      "RequestItems": Object {
        "DummyTable": Array [
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "0@email.com",
                },
                "id": Object {
                  "S": "0",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "1@email.com",
                },
                "id": Object {
                  "S": "1",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "2@email.com",
                },
                "id": Object {
                  "S": "2",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "3@email.com",
                },
                "id": Object {
                  "S": "3",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "4@email.com",
                },
                "id": Object {
                  "S": "4",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "5@email.com",
                },
                "id": Object {
                  "S": "5",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "6@email.com",
                },
                "id": Object {
                  "S": "6",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "7@email.com",
                },
                "id": Object {
                  "S": "7",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "8@email.com",
                },
                "id": Object {
                  "S": "8",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "9@email.com",
                },
                "id": Object {
                  "S": "9",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "10@email.com",
                },
                "id": Object {
                  "S": "10",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "11@email.com",
                },
                "id": Object {
                  "S": "11",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "12@email.com",
                },
                "id": Object {
                  "S": "12",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "13@email.com",
                },
                "id": Object {
                  "S": "13",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "14@email.com",
                },
                "id": Object {
                  "S": "14",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "15@email.com",
                },
                "id": Object {
                  "S": "15",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "16@email.com",
                },
                "id": Object {
                  "S": "16",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "17@email.com",
                },
                "id": Object {
                  "S": "17",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "18@email.com",
                },
                "id": Object {
                  "S": "18",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "19@email.com",
                },
                "id": Object {
                  "S": "19",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "20@email.com",
                },
                "id": Object {
                  "S": "20",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "21@email.com",
                },
                "id": Object {
                  "S": "21",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "22@email.com",
                },
                "id": Object {
                  "S": "22",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "23@email.com",
                },
                "id": Object {
                  "S": "23",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "24@email.com",
                },
                "id": Object {
                  "S": "24",
                },
              },
            },
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "RequestItems": Object {
        "DummyTable": Array [
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "25@email.com",
                },
                "id": Object {
                  "S": "25",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "26@email.com",
                },
                "id": Object {
                  "S": "26",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "27@email.com",
                },
                "id": Object {
                  "S": "27",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "28@email.com",
                },
                "id": Object {
                  "S": "28",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "29@email.com",
                },
                "id": Object {
                  "S": "29",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "30@email.com",
                },
                "id": Object {
                  "S": "30",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "31@email.com",
                },
                "id": Object {
                  "S": "31",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "32@email.com",
                },
                "id": Object {
                  "S": "32",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "33@email.com",
                },
                "id": Object {
                  "S": "33",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "34@email.com",
                },
                "id": Object {
                  "S": "34",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "35@email.com",
                },
                "id": Object {
                  "S": "35",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "36@email.com",
                },
                "id": Object {
                  "S": "36",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "37@email.com",
                },
                "id": Object {
                  "S": "37",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "38@email.com",
                },
                "id": Object {
                  "S": "38",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "39@email.com",
                },
                "id": Object {
                  "S": "39",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "40@email.com",
                },
                "id": Object {
                  "S": "40",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "41@email.com",
                },
                "id": Object {
                  "S": "41",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "42@email.com",
                },
                "id": Object {
                  "S": "42",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "43@email.com",
                },
                "id": Object {
                  "S": "43",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "44@email.com",
                },
                "id": Object {
                  "S": "44",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "45@email.com",
                },
                "id": Object {
                  "S": "45",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "46@email.com",
                },
                "id": Object {
                  "S": "46",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "47@email.com",
                },
                "id": Object {
                  "S": "47",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "48@email.com",
                },
                "id": Object {
                  "S": "48",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "49@email.com",
                },
                "id": Object {
                  "S": "49",
                },
              },
            },
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "RequestItems": Object {
        "DummyTable": Array [
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "50@email.com",
                },
                "id": Object {
                  "S": "50",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "51@email.com",
                },
                "id": Object {
                  "S": "51",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "52@email.com",
                },
                "id": Object {
                  "S": "52",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "53@email.com",
                },
                "id": Object {
                  "S": "53",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "54@email.com",
                },
                "id": Object {
                  "S": "54",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "55@email.com",
                },
                "id": Object {
                  "S": "55",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "56@email.com",
                },
                "id": Object {
                  "S": "56",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "57@email.com",
                },
                "id": Object {
                  "S": "57",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "58@email.com",
                },
                "id": Object {
                  "S": "58",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "59@email.com",
                },
                "id": Object {
                  "S": "59",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "60@email.com",
                },
                "id": Object {
                  "S": "60",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "61@email.com",
                },
                "id": Object {
                  "S": "61",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "62@email.com",
                },
                "id": Object {
                  "S": "62",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "63@email.com",
                },
                "id": Object {
                  "S": "63",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "64@email.com",
                },
                "id": Object {
                  "S": "64",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "65@email.com",
                },
                "id": Object {
                  "S": "65",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "66@email.com",
                },
                "id": Object {
                  "S": "66",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "67@email.com",
                },
                "id": Object {
                  "S": "67",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "68@email.com",
                },
                "id": Object {
                  "S": "68",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "69@email.com",
                },
                "id": Object {
                  "S": "69",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "70@email.com",
                },
                "id": Object {
                  "S": "70",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "71@email.com",
                },
                "id": Object {
                  "S": "71",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "72@email.com",
                },
                "id": Object {
                  "S": "72",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "73@email.com",
                },
                "id": Object {
                  "S": "73",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "74@email.com",
                },
                "id": Object {
                  "S": "74",
                },
              },
            },
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "RequestItems": Object {
        "DummyTable": Array [
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "75@email.com",
                },
                "id": Object {
                  "S": "75",
                },
              },
            },
          },
          Object {
            "PutRequest": Object {
              "Item": Object {
                "email": Object {
                  "S": "76@email.com",
                },
                "id": Object {
                  "S": "76",
                },
              },
            },
          },
        ],
      },
    },
  ],
]
`;
